File: fusefs/src/crypto-v1.js
const crypto = require('node:crypto');

// v1 crypto primitives and file formats.
//
// Design goals:
// - Use Node's built-in, well-vetted primitives (no custom crypto).
// - Simple versioned formats with authenticated headers.
// - AEAD for confidentiality + integrity.
//
// v1 chooses AES-256-GCM because it's widely supported by Node's crypto.
// (The design doc previously mentioned XChaCha20-Poly1305; we intentionally
// pick a primitive that is guaranteed available without native deps.)

const V1 = {
  FILE_MAGIC: Buffer.from('OCFS1', 'utf8'),
  FILE_VERSION: 0x01,
  ALG_AES_256_GCM: 0x01,
  NONCE_LEN: 12, // recommended size for GCM
  TAG_LEN: 16,
};

function randomKey32() {
  return crypto.randomBytes(32);
}

function sealAes256Gcm({ key, nonce, aad, plaintext }) {
  if (!Buffer.isBuffer(key) || key.length !== 32) throw new Error('key must be 32 bytes');
  if (!Buffer.isBuffer(nonce) || nonce.length !== V1.NONCE_LEN) throw new Error('nonce must be 12 bytes');
  if (!Buffer.isBuffer(plaintext)) throw new Error('plaintext must be a Buffer');

  const cipher = crypto.createCipheriv('aes-256-gcm', key, nonce);
  if (aad && aad.length) cipher.setAAD(aad);

  const ciphertext = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  const tag = cipher.getAuthTag();
  if (tag.length !== V1.TAG_LEN) throw new Error('unexpected auth tag length');

  return { ciphertext, tag };
}

function openAes256Gcm({ key, nonce, aad, ciphertext, tag }) {
  if (!Buffer.isBuffer(key) || key.length !== 32) throw new Error('key must be 32 bytes');
  if (!Buffer.isBuffer(nonce) || nonce.length !== V1.NONCE_LEN) throw new Error('nonce must be 12 bytes');
  if (!Buffer.isBuffer(ciphertext)) throw new Error('ciphertext must be a Buffer');
  if (!Buffer.isBuffer(tag) || tag.length !== V1.TAG_LEN) throw new Error('tag must be 16 bytes');

  const decipher = crypto.createDecipheriv('aes-256-gcm', key, nonce);
  if (aad && aad.length) decipher.setAAD(aad);
  decipher.setAuthTag(tag);

  // crypto throws on auth failure.
  return Buffer.concat([decipher.update(ciphertext), decipher.final()]);
}

function encodeEncryptedFileV1({ dek, plaintext }) {
  if (!Buffer.isBuffer(dek) || dek.length !== 32) throw new Error('dek must be 32 bytes');
  if (!Buffer.isBuffer(plaintext)) throw new Error('plaintext must be a Buffer');

  const nonce = crypto.randomBytes(V1.NONCE_LEN);
  const header = Buffer.concat([
    V1.FILE_MAGIC,
    Buffer.from([V1.FILE_VERSION, V1.ALG_AES_256_GCM, V1.NONCE_LEN]),
    nonce,
  ]);

  const { ciphertext, tag } = sealAes256Gcm({ key: dek, nonce, aad: header, plaintext });
  return Buffer.concat([header, ciphertext, tag]);
}

function decodeEncryptedFileV1({ dek, blob }) {
  if (!Buffer.isBuffer(dek) || dek.length !== 32) throw new Error('dek must be 32 bytes');
  if (!Buffer.isBuffer(blob)) throw new Error('blob must be a Buffer');

  const minLen = V1.FILE_MAGIC.length + 3 + V1.NONCE_LEN + V1.TAG_LEN;
  if (blob.length < minLen) throw new Error('ciphertext blob too small');

  const magic = blob.subarray(0, V1.FILE_MAGIC.length);
  if (!magic.equals(V1.FILE_MAGIC)) throw new Error('bad magic');

  const version = blob[V1.FILE_MAGIC.length + 0];
  const alg = blob[V1.FILE_MAGIC.length + 1];
  const nonceLen = blob[V1.FILE_MAGIC.length + 2];

  if (version !== V1.FILE_VERSION) throw new Error(`unsupported version: ${version}`);
  if (alg !== V1.ALG_AES_256_GCM) throw new Error(`unsupported alg: ${alg}`);
  if (nonceLen !== V1.NONCE_LEN) throw new Error(`unsupported nonceLen: ${nonceLen}`);

  const headerLen = V1.FILE_MAGIC.length + 3 + V1.NONCE_LEN;
  const header = blob.subarray(0, headerLen);
  const nonce = blob.subarray(V1.FILE_MAGIC.length + 3, headerLen);

  const tag = blob.subarray(blob.length - V1.TAG_LEN);
  const ciphertext = blob.subarray(headerLen, blob.length - V1.TAG_LEN);

  return openAes256Gcm({ key: dek, nonce, aad: header, ciphertext, tag });
}

module.exports = {
  V1,
  randomKey32,
  sealAes256Gcm,
  openAes256Gcm,
  encodeEncryptedFileV1,
  decodeEncryptedFileV1,
};

File: fusefs-swift/Sources/OcProtectFsFuse/CryptoV1.swift
import Foundation
import CryptoKit

// Phase 3 (Issue #109): crypto/policy/authz port.
// This file starts by porting the on-disk encrypted file format used by the Node implementation.
//
// Format (v1):
//   magic "OCFS1" (5 bytes)
//   version (1 byte) = 0x01
//   alg (1 byte) = 0x01 (AES-256-GCM)
//   nonceLen (1 byte) = 12
//   nonce (12 bytes)
//   ciphertext (...)
//   tag (16 bytes)
//
// AEAD AAD = header (magic+version+alg+nonceLen+nonce)

enum OcfsCrypto {
  enum Initial {
    static let fileMagic = Data("OCFS1".utf8)
    static let fileVersion: UInt8 = 0x01
    static let algAes256Gcm: UInt8 = 0x01
    static let nonceLen = 12
    static let tagLen = 16
  }

  struct DecodeError: Error, CustomStringConvertible {
    let description: String
  }

  /// Encode an encrypted file blob (Node-compatible).
  ///
  /// - Parameters:
  ///   - dek: 32-byte data encryption key
  ///   - plaintext: file contents
  ///   - nonce: optional 12-byte nonce (for tests). If nil, a random nonce is generated.
  static func encodeEncryptedFileV1(dek: Data, plaintext: Data, nonce: Data? = nil) throws -> Data {
    guard dek.count == 32 else { throw DecodeError(description: "dek must be 32 bytes") }

    let useNonce: Data
    if let nonce {
      guard nonce.count == Initial.nonceLen else { throw DecodeError(description: "nonce must be 12 bytes") }
      useNonce = nonce
    } else {
      var b = [UInt8](repeating: 0, count: Initial.nonceLen)
      _ = SecRandomCopyBytes(kSecRandomDefault, b.count, &b)
      useNonce = Data(b)
    }

    var header = Data()
    header.append(Initial.fileMagic)
    header.append(contentsOf: [Initial.fileVersion, Initial.algAes256Gcm, UInt8(Initial.nonceLen)])
    header.append(useNonce)

    let key = SymmetricKey(data: dek)
    let nonce = try AES.GCM.Nonce(data: useNonce)

    let sealed = try AES.GCM.seal(plaintext, using: key, nonce: nonce, authenticating: header)

    guard sealed.tag.count == Initial.tagLen else {
      throw DecodeError(description: "unexpected auth tag length")
    }

    var out = Data()
    out.append(header)
    out.append(sealed.ciphertext)
    out.append(sealed.tag)
    return out
  }

  /// Decode a v1 encrypted blob (Node-compatible) and return plaintext.
  static func decodeEncryptedFileV1(dek: Data, blob: Data) throws -> Data {
    guard dek.count == 32 else { throw DecodeError(description: "dek must be 32 bytes") }

    let minLen = Initial.fileMagic.count + 3 + Initial.nonceLen + Initial.tagLen
    guard blob.count >= minLen else { throw DecodeError(description: "ciphertext blob too small") }

    let magic = blob.subdata(in: 0..<Initial.fileMagic.count)
    guard magic == Initial.fileMagic else { throw DecodeError(description: "bad magic") }

    let version = blob[Initial.fileMagic.count + 0]
    let alg = blob[Initial.fileMagic.count + 1]
    let nonceLen = Int(blob[Initial.fileMagic.count + 2])

    guard version == Initial.fileVersion else { throw DecodeError(description: "unsupported version: \(version)") }
    guard alg == Initial.algAes256Gcm else { throw DecodeError(description: "unsupported alg: \(alg)") }
    guard nonceLen == Initial.nonceLen else { throw DecodeError(description: "unsupported nonceLen: \(nonceLen)") }

    let headerLen = Initial.fileMagic.count + 3 + Initial.nonceLen
    let header = blob.subdata(in: 0..<headerLen)
    let nonceData = blob.subdata(in: (Initial.fileMagic.count + 3)..<headerLen)

    let tagStart = blob.count - Initial.tagLen
    let ciphertext = blob.subdata(in: headerLen..<tagStart)
    let tag = blob.subdata(in: tagStart..<blob.count)

    let key = SymmetricKey(data: dek)
    let nonce = try AES.GCM.Nonce(data: nonceData)

    let sealed = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag)
    return try AES.GCM.open(sealed, using: key, authenticating: header)
  }
}
